### ISSUE TITLE
Create a new project

### ISSUE THREAD

**Issue by JMurphyWeb** [karmaradio/karma#4]
As a project manager, I can create a new project

Fields:
type **String** - (dropdown: feature film, television) 
budget **String** - (dropdown: no/low, mid, big) 
name (film title) **String**
codename (a 'known-as' title for confidentiality) **String**
description **Text**
start_date **Ecto.Date**
duration (weeks) **Integer**
studio_name **String**
company_name **String**
company_address_1 **String**
company_address_2 **String** _NOT REQUIRED_
company_address_city **String**
company_address_postcode **String**
company_address_county **String**

operating_base_address_1 **String** 
operating_base_address_2 **String**  _NOT REQUIRED_
operating_base_address_city **String** 
operating_base_address_postcode **String** 
operating_base_address_country **String** 

locations **String**
holiday_rate **Float** (dropdown. percentage: 10.77%, or 12.07%)
additional_notes **Text**

_Extra (separate table)_
file uploads (2 for now)  **String (URL)**  #9 

Questions:

- [x] can you give an example of what the project manager might want to enter for "locations"?
- [x] what files will the PM be uploading here for the contractor to be able to see when they receive an offer to work on this project?
- [x] Is there anything else?

**Comment by JMurphyWeb** [comment]
@jonscottfilm replied with the following in this https://github.com/karmaradio/karma/issues/8#issuecomment-299872739

Locations will be postal addresses of locations selected for shooting. These are unlikely to be known or fixed at the offer stage.

Project Manager will decide Department at the offer stage, and should decide job title...

Only files PM will upload will be the custom forms which each studio will provide for each project. Forms like Company Policies: Health & Safety, Security etc. Stuff each user should view, prior to signing some of them, via Webmerge & Docusign.

**Comment by jonscottfilm** [comment]
I know you'll know this, and you may already have it in hand, but I'm conscious that field-names need to be _very_ specific, because there will be thousands ultimately. eg. 'start_date' typically refers to an individual's work start date, so for a project should be 'project_start_date'

All fields for 'create new project'. 4 new one's in **bold**.

type (dropdown)
budget (dropdown)
name _(film title)_
**codename** _(a 'known-as' title for confidentiality, e.g. Dr Strange was Checkmate)_ 
description
start_date
duration (weeks)
**studio_name**
company_name
company_address
**operating_base_address** _(might be different from company address, eg. registered company is Central London, base is Shepperton Studios)_
locations
**holiday_rate** _(dropdown. percentage: 10.77, or 12.07%. Different studios handle it differently)_
additional notes
file upload

**Comment by JMurphyWeb** [comment]
This is useful, thanks. We need to establish the following:
- what is the content for the drop downs (I understand these are listed in a separate doc, but we should add them into this issue when finalised)
- any address listed -> is 4 fields enough?
- locations -> what are they likely to be? Do we need multiple fields or can they just be one field with the user putting commas between locations?
- how many file uploads will there be at this stage? is there an upper limit?

**Comment by jonscottfilm** [comment]
**will add dropdown content to this issue**

any address listed -> is 4 fields enough? 
**should be, but it never is**

locations -> what are they likely to be? Do we need multiple fields or can they just be one field with the user putting commas between locations?
**Long-term these will be full addresses, which will then play their part in features, such as timecard site of working records, mileage between work sites, and in fact agreements with the location operator themselves**
**For now, I _think_ this is just for information purposes, if know, i.e., this project shoots in London, Paris, Dubai and Shepperton Studios**

how many file uploads will there be at this stage? is there an upper limit?
**there could be up to 20 I'd say. It can get ridiculous.**

**Comment by JMurphyWeb** [comment]
@jonscottfilm I am going to look into make a start on this feature. I have copied your previous comment into the original issue. It's worth you taking a look to just confirm that these are everything we need at this stage

**Comment by JMurphyWeb** [comment]
I used the following generator here

> user_id:references:users
type:string 
budget:string 
name:string
codename:string
description:text
start_date:datetime
duration:integer
studio_name:string
company_name:string
company_address_1:string
company_address_2:string
company_address_3:string
company_address_4:string
company_address_5:string
operating_base_address_1:string
operating_base_address_2:string
operating_base_address_3:string
operating_base_address_4:string
operating_base_address_5:string
locations:string
holiday_rate:float
additional_notes:text
active:boolean

**Comment by JMurphyWeb** [comment]
My tasks for this PR #35 

- [ ] Run generator
- [ ] fix up migration file
- [ ] fix up model
- [ ] ensure users can only view their own projects

Future:
- [ ] Ensure users cannot view other peoples projects

**Comment by jonscottfilm** [comment]
I think this list of fields is finalised.

Only​ comment I would make, and you may have it in hand, is about fieldnames being very specific. Eg, company_name will appear similar to contractor's company name, or vendor profile (non-mvp) company name... So, does it need to be project_company_name..? 
Likewise should start_date be project_start_date...? Or not?

**Comment by jonscottfilm** [comment]
Thought you might say that. Perfect!

**Comment by JMurphyWeb** [comment]
@jonscottfilm I am using the 'multiple fields' format for all addresses: https://www.gov.uk/service-manual/design/addresses#using-multiple-fields

**Comment by JMurphyWeb** [comment]
The basic structure is in. Tasks for today:

- [ ] Convert generated tests to make more meaningful & readable
- [ ] Add permissions for show, edit, update, delete project
- [ ] Look into adding some seed data

Follow on tasks not directly related to current PR/issue
- [ ] Dashboard view should load users projects #19 
- [ ] Build basic project dashboard

**Comment by JMurphyWeb** [comment]
@jonscottfilm we were thinking of adding a 'save draft' option in the create project view as there are quite a few fields. 

Question is, which fields should we require in order to save a draft? 

Also in order to submit a new project (not in draft form) all fields from the original comment https://github.com/karmaradio/karma/issues/4#issue-227043925 are required apart from the `..._address_2` ones.

Any comments/questions on that?

**Comment by jonscottfilm** [comment]
Save draft project. Um, I hadn't really considered it. Just create project, then save it. I guess because it will be quite a rare process, one per project! As opposed to draft offers, which will be hundreds of offers per project.
Maybe just have say 3 essential fields before giving the ability to save a project? Project name, codename, studio, company?

**Comment by jonscottfilm** [comment]
Correct re all fields in #4 (comment). Not including 2nd lines of address - is that because you have multi-line entry in 1st line address?

**Comment by JMurphyWeb** [comment]
Above were discussed on call. Save draft will be left for now, and address line 2 is not a required field in order to create a new project.

___________________________

I am now going to look into adding some styling

**Comment by jonscottfilm** [comment]
Minor point, but 'Codename' should be one word. Can be written as two words, but not hyphenated.

Start Date field should be date-picker.
Project date picker should default to TODAY.

Error fields need to be highlighted to be found and corrected.

**Comment by jonscottfilm** [comment]
Start Date field needs to specify '**_shoot_** start date' , or Start Date (of shoot).
Maybe in helper text underneath field (later sprint), saying:
_'approximate start date of shooting'_ 

Duration field throws an error. It allows user to input text, but it only likes numbers. I tested it by entering 10 weeks. It accepted it when corrected to 10 only. 
Long-terms, might want to change to number and week/month chooser? PM might refer to shoot period as 4 months?

**Comment by jonscottfilm** [comment]
'Name' (project name) field, in Chrome is autofilled by 'Name', eg:

![image](https://cloud.githubusercontent.com/assets/28431542/26267556/bca10f80-3ce2-11e7-84d0-0101a828b657.png)

**Comment by JMurphyWeb** [comment]
Working on the name autocorrect and the start date label/instruction now.

Would you agree that limiting the user to only be able to enter numbers for the duration is an ok solution for now?

**Comment by jonscottfilm** [comment]
Another instance of browser auto-fill errror/mis-match, is Box Rental Cap, in Create Offer flow:
![image](https://cloud.githubusercontent.com/assets/28431542/26304416/218ff24c-3ee3-11e7-841f-068d8da43105.png)

**Comment by jonscottfilm** [comment]
Numbers only is fine. 
Same to apply to all number fields? (looks like you might already be doing that?) 
And can it trigger numberpad input on mobile devices?

**Comment by JMurphyWeb** [comment]
Yes, number keyboard appears when clicking a 'number' input.

Yup should already be doing that.

Start date input label has been updated.

Code-name updated to Codename

All fields with errors should be appended with a prompt. Please let me know if you find one which doesn't.

Not quite sure I understand the datepicker point, you want all inputs of type date to be date pickers? If so we can make a new issue for that (more of an enhancement than bug)

Changes aren't merged in just yet, will re-assign you when it is ready.

**Comment by JMurphyWeb** [comment]
I need to fix/look into autofill

**Comment by jonscottfilm** [comment]
Might just be a hangover from copy/pasting the code from your form field template(?) of eg the 'Name' field', and changing it to eg Box Rental... ??

Or, it's my aggressive Chrome autofill!

**Comment by jonscottfilm** [comment]
Did we settle on minimum number of / specific fields required in order to create a project?

**Comment by JMurphyWeb** [comment]
I think we are now not autofilling by default. Hard for me to test as chrome wasn't being quite as forceful with me. Again once you are assigned you can check whether or not it is done.

**Comment by JMurphyWeb** [comment]
sorry for the copy paste job, but these are the currently required projects:

```elixir
      :type,
      :budget,
      :name,
      :codename,
      :description,
      :start_date,
      :duration,
      :studio_name,
      :company_name,
      :company_address_1,
      :company_address_city,
      :company_address_postcode,
      :company_address_country,
      :operating_base_address_1,
      :operating_base_address_city,
      :operating_base_address_postcode,
      :operating_base_address_country,
      :holiday_rate,
      :active,
      :user_id
```

**Comment by jonscottfilm** [comment]
Be back in action at 3.30. Will test then.

However, we DO like autocomplete, so can it come back to play later....?! Makes life easier for users.

OK, you remember at the point you were considering 'draft projects' we talked about approx 3 to 5 fields before being able to save a project? 
Not all 20 of the above may be known, eg operating base, at point of project creation.
I can PM's having issue answering Budget, too.

**Comment by JMurphyWeb** [comment]
Autocomplete can be added for specific fields. Should still will work for login for example.

**Comment by JMurphyWeb** [comment]
I think draft projects were shelved for now. Regardless, it would require a separate issue with the minimum required fields. Think there would need to be a discussion on implementation too.

**Comment by jonscottfilm** [comment]
Forget drafts.
Will just make some fields not required.

Auto-complete not working for log-in. And password is shown, not obscured with ***

**Comment by jonscottfilm** [comment]
Login bug fixed. 

We can lose 'eg 16 weeks' in the 'Duration' field. Thx

Then this can be closed.

**Comment by jackcarlisle** [comment]
@jonscottfilm what should the placeholder be replaced with? Or just leave blank?

**Comment by jonscottfilm** [comment]
Sorry, I guess just 'Enter a number', like the offer field placeholders. I appreciate it is helpful, but maybe too much!

Could lose the fullstop after Company Ltd. in Company Name if you're in there, or change to Limited if that is signifying the abbreviation?

**Comment by jackcarlisle** [comment]
yeah sure!

**Comment by jonscottfilm** [comment]
In fact, the labels and placeholders have slight inconsistency. Shall we go all sentence case, like the labels format, ie. 'Operating base address 2'

**Comment by jackcarlisle** [comment]
yeah I'll change that too

**Comment by jonscottfilm** [comment]
Not priority, but I guess at some point we will need admins option to **delete projects**.

(just I'm getting kinda crowded, and then highlights to me that the (requested by me) 'Add Projects' button, is way out of sight... :))

![image](https://cloud.githubusercontent.com/assets/28431542/26318911/ba81a4a6-3f14-11e7-8ca3-2b599c667882.png)

**Comment by jackcarlisle** [comment]
@jonscottfilm yeah we will add that functionality. Would you prefer if the button was at the top of the list instead?

**Comment by jonscottfilm** [comment]
Let's try the 'Add project' button at top right of list, in line with My projects.

I wonder, as & when we get to it, if the Delete project button should be hidden away at the bottom of Edit project view?

**Comment by jackcarlisle** [comment]
Ok changing now.

Yeah that might be a good place to put it to avoid accidentally pressing it. (We'll have checks in place asking if they mean to delete it before doing so)

**Comment by jonscottfilm** [comment]
👍

**Comment by jackcarlisle** [comment]
<img width="1183" alt="screen shot 2017-05-22 at 17 51 10" src="https://cloud.githubusercontent.com/assets/12450298/26319531/4e2558b8-3f17-11e7-8149-ceb329179caa.png">

**Comment by jonscottfilm** [comment]
Excellent. Any chance it can be smaller please :)

**Comment by jackcarlisle** [comment]
yeah sure!

**Comment by jackcarlisle** [comment]
<img width="1187" alt="screen shot 2017-05-22 at 17 56 18" src="https://cloud.githubusercontent.com/assets/12450298/26319716/fd3a0c40-3f17-11e7-81a4-f4072ceafc65.png">

**Comment by jonscottfilm** [comment]
Beauty. Thanks!

**Comment by jonscottfilm** [comment]
Start Date label needs to read either..

**Shoot Start Date**
or
**Start Date (of shoot)**

..to differentiate between other project milestone dates which PM will be aware of, and although not entering now, will play in Karma later verrsions, i.e:

Pre-Pre-Production Start Date
Pre-Production Start Date
Post-Production Start Date
Picture-Lock Date
Release Date

**Comment by jonscottfilm** [comment]
Only other comment would be perhaps re the 'Studio name' label. It just needs a more generic term in the label, I think. For those with no studio, or a closer parent company not considered a studio, eg 'Plan-B' or 'Working Title'.

So maybe revise label to: **Studio/Production Company**

As & when :)

**Comment by JMurphyWeb** [comment]
ready for testing.

**Comment by jonscottfilm** [comment]
Did you clock this comment:

'Start Date label needs to read either..

Shoot Start Date
or
Start Date (of shoot)

..to differentiate between other project milestone dates which PM will be aware of, and although not entering now, will play in Karma later verrsions, i.e:

Pre-Pre-Production Start Date
Pre-Production Start Date
Post-Production Start Date
Picture-Lock Date
Release Date'

**Comment by jonscottfilm** [comment]
Amend label, per https://github.com/karmaradio/karma/issues/4#issuecomment-303733372
and this can be closed.

**Comment by JMurphyWeb** [comment]
will do

**Comment by JMurphyWeb** [comment]
@jonscottfilm this is now done

**Comment by jonscottfilm** [comment]
Working. Closing.
